src/System.Private.ServiceModel/tests/Common/Infrastructure/src/ConditionalWcfTest.cs

```diff
@@ -10,42 +15,191 @@ namespace Infrastructure.Common
+        // 'true' or 'false', it will be used as is.  If the value is
+        // empty, and 'detectFunc' is specified, it will be invoked to
+        // evaluate the condition at runtime.  The value is cached, so
+        // 'detectFunc' will be called only once.
+        private static bool GetConditionValueFromTestProperties(string conditionName, Func<bool> detectFunc = null)
+        {
+            string propertyValue;
+            return GetConditionValue(conditionName, () =>
+            {
+                propertyValue = TestProperties.GetProperty(conditionName);
+                if (string.IsNullOrWhiteSpace(propertyValue))
+                {
+                    // Empty string means auto-detect if possible, else false
+                    return (detectFunc == null) ? false : detectFunc();
+                }
+                else
+                {
+                    bool conditionResult = false;
+                    if (!bool.TryParse(propertyValue, out conditionResult))
+                    {
+                        conditionResult = false;
+                    }
+
+                    return conditionResult;
+                }
+            });
+        }
+
+        private static bool Detect_Server_Is_Local()
+        {
+            // Temporary workaround to detect whether this test is using localhost
+            // for either new or old test infrastructure.
+            // Refactor this after integration to address https://github.com/dotnet/wcf/issues/1024 
+            string host = null;
+            if (TestProperties.PropertyNames.Contains("ServiceUri"))
+            {
+                host = TestProperties.GetProperty("ServiceUri");
+            }
+            else if (TestProperties.PropertyNames.Contains("BridgeHost"))
+            {
+                host = TestProperties.GetProperty("BridgeHost");
+            }
+             if (String.IsNullOrWhiteSpace(host))
+            {
+                return false;
+            }
+            int index = host.IndexOf("localhost", 0, StringComparison.OrdinalIgnoreCase);
+            return index >= 0;
+        }
+        
+        private static bool Detect_Is_Windows()
+        {
+            return RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
         }
 
         public static bool Domain_Joined()
         {
-            return IsConditionalTestEnabled(TestProperties.Domain_Joined_PropertyName);
+            // If TestProperties does not specify true/false explicitly, we need to autodetect whether
+            // the client and server are domain joined.
+            // Temporarily use the simple heuristic that if we are running the services locally, it is.
+            // Refactor this after integration to address https://github.com/dotnet/wcf/issues/1024 
+            return GetConditionValueFromTestProperties(TestProperties.Domain_Joined_PropertyName,
+                                                       Detect_Server_Is_Local);
```

Comment:
> I'm under the assumption that our current [ConditionalFact] discussion is based only on things that can be entirely decided on the client

ConditionalFact is not limited to local conditions. It's our auto detection that can tell only local conditions for the most of the time.

>  I would prefer to see methods like "Is_Service_Local" and "Is_Domain_Joined" be in one place and re-used by any code that needs it.

So sounds like the condition you really mean is something like "Is_Client_Domain_Joined". For tests that will work only if both client and remote server are domain joined, they will need an additional condition, something like "Is_Server_Domain_Joined". The condition "Is_Server_Domain_Joined" can potentially be auto detected if "Is_Service_Local" is true.

> My interpretation of this initial checkin was to become concrete about all the conditions we need to branch on, and then to provide solid implementations later.

Agreed. I think current discussion _is_ about what are concrete conditions for domain joined scenarios. I did not bother implementation too much, although I prefer to see most auto detection logic to be more restrict (return false if cannot be sure).